### PR TITLE
Trigger post hook on Event Copy

### PR DESCRIPTION
**Pull request by mattwire**
Overview
----------------------------------------
When creating a copy of an event via API Event.create (with a template ID) (or directly using the CRM_Event_BAO_Event::copy function - as called from ManageEvents-"Copy" in the UI) a "copy" hook is called but the expected Event.create hook is not called, despite a new event being created.

In addition the copy function (which is only called in two places) has complexities to support passing in an already copied event.  This seems unnecessarily complex and this PR simplifies that so it is always responsible for creating the copied event.

Note: Copying events via the "Repeat Events" (RecurringEntity) functionality doesn't use this function... It hardcodes it's own version which should be replaced with this copy function in a future PR.

Before
----------------------------------------
`CRM_Event_BAO_Event::copy()` doesn't call the "create" hook when an event is created and adds unnecessary complexity by allowing an already created event to be passed in.

After
----------------------------------------
`CRM_Event_BAO_Event::copy()` calls the "create" (post) hook when an event is created and simplifies the function so it is easier to maintain and simpler.

Technical Details
----------------------------------------
The "pre" hook is still not implemented because this would require conversion from an object to a params array and changes can be made via the "post" hook if required.

Comments
----------------------------------------
This change should be NFC.  The API remains the same, if extensions are calling `CRM_Event_BAO_Event::copy()` directly (which is unsupported) they will need to switch to use the API. However I'm not aware of any extensions that do call that function.

### CODE REVIEW

File: CRM/Event/BAO/Event.php [hunk 1]
```diff
@@ -944,43 +942,37 @@ public static function copy($id, $newEvent = NULL, $afterCreate = FALSE) {
...
-    if ($newEvent && is_a($newEvent, 'CRM_Event_DAO_Event')) {
-      $copyEvent = $newEvent;
-    }
-
-    if (!isset($copyEvent)) {
-      $copyEvent = &CRM_Core_DAO::copyGeneric('CRM_Event_DAO_Event',
-        array('id' => $id),
-        array(
-          'loc_block_id' =>
+    $copyEvent = &CRM_Core_DAO::copyGeneric('CRM_Event_DAO_Event',
```
Comment (eileenmcnaughton) [inline CRM/Event/BAO/Event.php]: this & is a php 4 construct - it can go if you are touching this code

**Comment by eileenmcnaughton** [comment]
@mattwire there is a style warning.

Could the hooks go in copyGeneric?

File: CRM/Event/BAO/Event.php (hunk 1 above)
Comment (mattwire) [reply CRM/Event/BAO/Event.php]: Yes

**Comment by mattwire** [comment]
> Could the hooks go in copyGeneric?

Yes. Much better, then it triggers on all the entities that are copied.

File: CRM/Export/Form/Select.php [hunk 2]
```diff
@@ -98,6 +98,7 @@ public function preProcess() {
     $components = array('Contact', 'Contribute', 'Member', 'Event', 'Pledge', 'Case', 'Grant', 'Activity');
 
     // FIXME: This should use a modified version of CRM_Contact_Form_Search::getModeValue but it doesn't have all the contexts
+    // FIXME: Or better still, use CRM_Core_DAO_AllCoreTables::getBriefName($daoName) to get the $entityShortName
```
Comment (mattwire) [inline CRM/Export/Form/Select.php]: Ok, this is comment is unrelated to this PR... but I was looking for a function to get the Entity shortname and I found it!

**Comment by mattwire** [comment]
test this please

**Review by jitendrapurohit — APPROVED** [review]
Tested this and confirmed the call of post hook on Event copy as well as while creating a copy of a Contribution page.

Also tested an API call with `$params['template_id']` and it copies the event correctly.

ok to merge.

**Comment by eileenmcnaughton** [comment]
thanks @jitendrapurohit merged. @mattwire - one comment for possible follow up above^^

**Comment by eileenmcnaughton** [comment]
@mattwire is there any related ticket to close?
